.github/labeler.yml

```diff
@@ -0,0 +1,144 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+# Pull Request Labeler Github Action Configuration: https://github.com/marketplace/actions/labeler
+#
+# Note that we currently cannot use the negatioon operator  (i.e. `!`)  for miniglob matches as they
+# would match any file that doesn't touch them. What's needed is the concept of `any `, which takes a
+# list of constraints / globs and then matches all of the constraints for either `any` of the files or
+# `all` of the files.
+#
+# However, `any`/`all` are not supported in a released version and testing off of the `main` branch
+# had them not working proprly. While we wait for this issue to be handled upstream, we can remove
+# the negative / not matches for now and at least have labels again.
+INFRA:
+  - ".github/**/*"
+  - "appveyor.yml"
+  - "tools/**/*"
+  - "dev/create-release/**/*"
+  - ".asf.yaml"
+  - ".gitattributes"
+  - ".gitignore"
+  - "dev/github_jira_sync.py"
+  - "dev/merge_spark_pr.py"
+  - "dev/run-tests-jenkins*"
+BUILD:
+ # Can be supported when a stable release with correct all/any is released
+ #- any: ['dev/**/*', '!dev/github_jira_sync.py', '!dev/merge_spark_pr.py', '!dev/.rat-excludes']
+ - "dev/**/*"
+ - "build/**/*"
+ - "project/**/*"
+ - "assembly/**/*"
+ - "**/*pom.xml"
+ - "bin/docker-image-tool.sh"
+ - "bin/find-spark-home*"
+ - "scalastyle-config.xml"
+ # These can be added in the above `any` clause (and the /dev/**/* glob removed) when
+ # `any`/`all` support is released
+ # - "!dev/github_jira_sync.py"
+ # - "!dev/merge_spark_pr.py"
+ # - "!dev/run-tests-jenkins*"
+ # - "!dev/.rat-excludes"
+DOCS:
+  - "docs/**/*"
+  - "**/README.md"
+  - "**/CONTRIBUTING.md"
+EXAMPLES:
+  - "examples/**/*"
+  - "bin/run-example*"
+# CORE needs to be updated when all/any are released upstream.
+CORE:
+  # - any: ["core/**/*", "!**/*UI.scala", "!**/ui/**/*"] # If any file matches all of the globs defined in the list started by `any`, label is applied.
+  - "core/**/*"
```

Comment:
Along the same lines as the `STRUCTURED STREAMING` tag, I'm wondering if this one also might need a leading glob for subdirectories called core.

Reply:
There are only three possible matches that I can find, as listed below. There are only three and of those one of them is the root `core` directory. So let's look at the other two.

```bash
$ find . -type d -iname 'core'           
./resource-managers/kubernetes/core
./core
./sql/core
```

Looking at the files in /resource-managers/kubernetes/core, it doesn't feel like any of them are "core" - but I could be misinformed on what we consider `CORE`. The files underneath the kubernetes path are all related to k8s deployment and scheduling (and all of the things that go with that on k8s, such as persistent volume claims and persistent volume mounts, etc). While these do feel potentially like `core` behaviors, if this were the case I would expect the other cluster managers to also fall under this section.

As for `./sql/core`, this one I do believe should be a match. One of the files below it is `sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala`. Definitely I think that the main file defining the `Dataset` type is core.

I would propose that we add `sql/core/**/*` to the list of globs for `CORE` so as to match everything underneath `sql/core` but to still avoid the things underneath `resource-managers`.

Reply:
I've yet to take action on this one, but I'm leading towards adding an additional rule for `sql/core/**/*` so as to avoid the resource-managers/kubernetes/core folders while still picking up the "core" SQL files such as Dataset.scala etc.

Does anybody have any thoughts on whether or not we'd like the files underneath `./resource-managers/kubernetes/core` to be labeled as `CORE`? My gut tells me no, but I'm still pretty new here so I'll go with whatever you guys suggest.  cc @HyukjinKwon @dongjoon-hyun @RussellSpitzer @rdblue 

After that, I believe this PR is ready to be merged. We will need to open an issue for the `actions/labeler` team to cut a release with `any` / `all` support. Their current suggestion in a relatively old issue (linked in the main issue at https://issues.apache.org/jira/browse/SPARK-33282) is to use the `main` branch (renamed from `master`), but I had difficulty with getting matches using that and overall I'm not that excited about the idea of using a non-released version.

I'd prefer to open an issue to cut a release with support for `all` and `any` (we really only need `any` to handle the negated matches correctly), add a link to the issue to the labeler file, and then merge this and update it once said release has been cut.

Reply:
I intended only to have a match with the root `core` when I added. If it works in a different way, we should fix it.
I would prefer to keep `sql/core/**/*` as `SQL`. That's what I usually have seen from other PRs.

Also, let's focus on converting it to labeler action instead of adding/removing/changing other patterns here.

Reply:
Ah ok. I wasn't sure on the original intended behavior for a few of them but I agree let's keep it focused on converting it to the labeler action for now (other than the few bits that don't seem to be stable yet - though I'm going to give those another go now that I've gotten a lot more experience with this labeler).

TLDR: I will leave it as originally intended 👍